### PR TITLE
tikv/safeTS: obtain min-resolved-ts by store list

### PR DESCRIPTION
**Pull request by HuSharp**
### Background
Nowadays we have 2 api interface for obtaining min resolved ts

- /pd/api/v1/min-resolved-ts: obtain cluster's min resolved ts
- /pd/api/v1/min-resolved-ts/{store_id}: obtain each store's min resolved ts

There are 2 repair step in this pr:
1. We can use the `approach I` to get cluster's min resolved ts when @@txn_scope is global.
2. For client-go's updateSafeTS, we call `approach II` for each store, which is not necessary.
We can extend the `approach I` params to obtain specified store via a list to reduce the cost per call after https://github.com/tikv/pd/pull/6880 merged

#### step 1
judge @@txn_scope value which from the config:

- when find is global, we can get cluster-level resolved ts to avoid using kv grpc.
- when find is not global, we need to use kv grpc.

#### step 2
use `approach I`'s extension to get all stores.

### CODE REVIEW

**Comment by HuSharp** [comment]
@JmPotato @lhy1024 PTAL as well, thx!

File: integration_tests/go.mod [hunk 1]
```diff
@@ -112,7 +112,5 @@ require (
 
 replace (
 	github.com/go-ldap/ldap/v3 => github.com/YangKeao/ldap/v3 v3.4.5-0.20230421065457-369a3bab1117
-	github.com/pingcap/tidb => github.com/HuSharp/tidb v1.1.0-beta.0.20230726045237-a2b0085ad7c5
-	github.com/pingcap/tidb/parser => github.com/HuSharp/tidb/parser v0.0.0-20230726045237-a2b0085ad7c5
```
Comment (HuSharp) [inline integration_tests/go.mod:116]: FIY, here update is independent of this pr, just for the other merged pr which wants to pass ci...

File: tikv/kv.go [hunk 2]
```diff
@@ -584,6 +584,23 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
 	tikvClient := s.GetTiKVClient()
 	wg := &sync.WaitGroup{}
 	wg.Add(len(stores))
+	// Try to get the minimum resolved timestamp of the store from PD.
+	var (
+		err error
+	)
+	var storeIDs []string
```
Comment (JmPotato) [inline tikv/kv.go:591]:
```suggestion
	var (
		storeIDs []string
		err           error
	)
```

File: tikv/kv.go [hunk 3]
```diff
@@ -595,16 +612,11 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
 
 			var (
 				safeTS uint64
-				err    error
 			)
```
Comment (JmPotato) [inline tikv/kv.go:615]: Please use `var safeTS uint64` if there is only one variable to init.

File: tikv/kv.go [hunk 4]
```diff
@@ -595,16 +612,11 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
...
-			storeIDStr := strconv.Itoa(int(storeID))
-			// Try to get the minimum resolved timestamp of the store from PD.
 			if s.pdHttpClient != nil {
```
Comment (JmPotato) [inline tikv/kv.go:616]: No need to do this check again.

File: tikv/kv.go [hunk 5]
```diff
@@ -595,16 +612,11 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
...
-				safeTS, err = s.pdHttpClient.GetStoreMinResolvedTS(ctx, storeID)
-				if err != nil {
-					logutil.BgLogger().Debug("get resolved TS from PD failed", zap.Error(err), zap.Uint64("store-id", storeID))
-				}
+				safeTS = StoreMinResolvedTSs[storeID]
 			}
+			storeIDStr := strconv.FormatUint(storeID, 10)
```
Comment (JmPotato) [inline tikv/kv.go:619]: What about moving this into the `if` block below?

Comment (HuSharp) [reply tikv/kv.go:619]: - metrics need string type to report... so need to at if block outer
- StoreMinResolvedTSs need unit type, so we cannot put it in func's params :(

File: .github/workflows/integration.yml [hunk 6]
```diff
@@ -69,7 +69,10 @@ jobs:
           mv ../${{steps.extract-tikv.outputs.destination}}/tikv-server .
           ./pd-server > pd.log 2>&1 &
           sleep 5
-          ./tikv-server -C tikv.toml > tikv.log 2>&1 &
+          TIKV_COUNT=3
+          for i in $(seq $TIKV_COUNT); do
+            ./tikv-server -C tikv.toml > "tikv${i}.log" 2>&1 &
```
Comment (JmPotato) [inline .github/workflows/integration.yml:74]: Will this cause the 3 TiKV nodes to use the same configuration? The listen address, data path, etc, may be conflicted.

File: tikv/kv.go [hunk 7]
```diff
@@ -584,6 +584,23 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
...
 	wg.Add(len(stores))
+	// Try to get the minimum resolved timestamp of the store from PD.
+	var (
+		storeIDs []string
+		err      error
+	)
+	for _, store := range stores {
+		storeIDs = append(storeIDs, strconv.FormatUint(store.StoreID(), 10))
+	}
+	StoreMinResolvedTSs := make(map[uint64]uint64)
```
Comment (JmPotato) [inline tikv/kv.go:595]: Why make it exported?

Comment (HuSharp) [reply tikv/kv.go:595]: for `for` block will use it.

File: .github/workflows/integration.yml (hunk 6 above)
Comment (HuSharp) [reply .github/workflows/integration.yml]: Oh, good catch! changed it for diff binding, PTAL again, thx!

File: tikv/kv.go (hunk 7 above)
Comment (JmPotato) [reply tikv/kv.go:595]: The exportable attribute is only meaningful when referenced across packages. The definition and use of `StoreMinResolvedTSs` are in the same function, so it is meaningless to make it exportable here.

Comment (HuSharp) [reply tikv/kv.go:595]: Sorry... I misunderstand your mean...

**Comment by HuSharp** [comment]
@nolouch PTAL, thx!

**Review by lhy1024 — APPROVED** [review]

**Comment by HuSharp** [comment]
friendly ping :) @JmPotato @nolouch

File: tikv/kv.go [hunk 8]
```diff
@@ -584,6 +584,21 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
 	tikvClient := s.GetTiKVClient()
 	wg := &sync.WaitGroup{}
 	wg.Add(len(stores))
+	// Try to get the minimum resolved timestamp of the store from PD.
+	var err error
+	var storeMinResolvedTSs map[uint64]uint64
```
Comment (JmPotato) [inline tikv/kv.go:589]:
```suggestion
	var (
		err error
		storeMinResolvedTSs map[uint64]uint64
	)
```

File: tikv/kv.go [hunk 9]
```diff
@@ -593,20 +608,10 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
...
-			storeIDStr := strconv.Itoa(int(storeID))
-			// Try to get the minimum resolved timestamp of the store from PD.
-			if s.pdHttpClient != nil {
-				safeTS, err = s.pdHttpClient.GetStoreMinResolvedTS(ctx, storeID)
-				if err != nil {
-					logutil.BgLogger().Debug("get resolved TS from PD failed", zap.Error(err), zap.Uint64("store-id", storeID))
-				}
-			}
+			var safeTS uint64
+			storeIDStr := strconv.FormatUint(storeID, 10)
```
Comment (JmPotato) [inline tikv/kv.go:612]: Could we reuse the results from lines 591~594?

File: util/pd.go [hunk 10]
```diff
@@ -86,45 +86,65 @@ func NewPDHTTPClient(
...
+						// Should be val.(uint64) but failpoint doesn't support that.
+						if tmp, ok := val.(int); ok {
+							d.StoresMinResolvedTS[storeID] = uint64(tmp)
+							logutil.BgLogger().Info("inject min resolved ts", zap.Uint64("storeID", storeID), zap.Uint64("ts", uint64(tmp)))
+						}
+					}
+				}
+			} else {
 				// Should be val.(uint64) but failpoint doesn't support that.
 				if tmp, ok := val.(int); ok {
```
Comment (JmPotato) [inline util/pd.go:133]:
```suggestion
			} else if tmp, ok := val.(int); ok {
				// Should be val.(uint64) but failpoint doesn't support that.
```

File: tikv/kv.go (hunk 9 above)
Comment (HuSharp) [reply tikv/kv.go:612]: good catch, reused by `for range index`

File: integration_tests/pd_api_test.go [hunk 11]
```diff
@@ -103,42 +105,107 @@ func (c *storeSafeTsMockClient) CloseAddr(addr string) error {
 	return c.Client.CloseAddr(addr)
 }
 
-func (s *apiTestSuite) TestGetStoreMinResolvedTS() {
+func (s *apiTestSuite) TestGetStoresMinResolvedTS() {
 	util.EnableFailpoints()
-	// Try to get the minimum resolved timestamp of the store from PD.
 	require := s.Require()
+	require.Nil(failpoint.Enable("tikvclient/mockFastSafeTSUpdater", `return()`))
```
Comment (JmPotato) [inline integration_tests/pd_api_test.go:111]:
```suggestion
	require.NoError(failpoint.Enable("tikvclient/mockFastSafeTSUpdater", `return()`))
```

File: integration_tests/pd_api_test.go [hunk 12]
```diff
@@ -103,42 +105,107 @@ func (c *storeSafeTsMockClient) CloseAddr(addr string) error {
...
+	defer func() {
+		require.Nil(failpoint.Disable("tikvclient/mockFastSafeTSUpdater"))
```
Comment (JmPotato) [inline integration_tests/pd_api_test.go:113]: Ditto.

File: tikv/kv.go [hunk 13]
```diff
@@ -655,6 +672,40 @@ func (s *KVStore) updateSafeTS(ctx context.Context) {
...
 }
 
+var (
+	skipSafeTSUpdateCounter    = metrics.TiKVSafeTSUpdateCounter.WithLabelValues("skip", "cluster")
+	successSafeTSUpdateCounter = metrics.TiKVSafeTSUpdateCounter.WithLabelValues("success", "cluster")
+	clusterMinSafeTSGap        = metrics.TiKVMinSafeTSGapSeconds.WithLabelValues("cluster")
+)
+
+// setClusterMinSafeTSByPD check whether it is needed to get cluster-level's min resolved ts from PD.
+func (s *KVStore) setClusterMinSafeTSByPD(ctx context.Context) bool {
```
Comment (JmPotato) [inline tikv/kv.go:682]: This function's name sounds like a method that will set something, which is not, however.

File: integration_tests/pd_api_test.go (hunk 11 above)
Comment (HuSharp) [reply integration_tests/pd_api_test.go]: seems that this usage in `client-go` needs to be unified

File: tikv/kv.go (hunk 13 above)
Comment (HuSharp) [reply tikv/kv.go]: updated, what do u think :)

**Review by JmPotato — APPROVED** [review]

**Pull request merged**